src/radical/pilot/agent/scheduler/base.py

```diff
@@ -310,14 +314,33 @@ def create(cls, cfg, session):
     #
     # NOTE: any scheduler implementation which uses a different nodelist
     #       structure MUST overload this method.
+
+    # MING: This function is used to update the node list by taking in a list
+    #       of slots that have been assigned an update them to the new state.
+    #       The nodes are represented as a dictionary, where as the slots are
+    #       represented as a dictionary of a different structure. This confused
+    #       me at first when I tried to read it.
+
```

Comment:
I can see that.  Can you add references to the documentation of both structures?  Those should be documented in the base class again.  Hmm, I wonder if that method should actually also be moved to the base class.  Any thoughts?

Reply:
It seems like the different implementations of schedulers (derived from the base class) change slot states differently. The continuous scheduler may use `_change_slot_states` from the base class, scattered scheduler uses the `_change_slot_states` from the scattered scheduler class, the yarn scheduler doesn't use `_change_slot_states` at all.

It might be better to push this function to the derived class because the logic to decide how to assign resources to nodes is in the derived classes anyways. Any thoughts?